nifi-nar-bundles/nifi-cdc/nifi-cdc-postgresql-bundle/nifi-cdc-postgresql-processors/src/main/java/org/apache/nifi/cdc/postgresql/processors/CaptureChangePostgreSQL.java

```diff
@@ -0,0 +1,468 @@
+        final String username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+        final boolean dropSlotIfExists = context.getProperty(DROP_SLOT_IF_EXISTS).evaluateAttributeExpressions().asBoolean();
+        final String publicationName = context.getProperty(PUBLICATION).evaluateAttributeExpressions().getValue();
+        final String replicationSlotName = context.getProperty(SLOT_NAME).evaluateAttributeExpressions().getValue();
+
+        // Save off PostgreSQL cluster and JDBC driver information, will be used to
+        // connect for event
+        final String driverName = context.getProperty(DRIVER_NAME).evaluateAttributeExpressions().getValue();
+        final String driverLocation = context.getProperty(DRIVER_LOCATION).evaluateAttributeExpressions().getValue();
+
+        String password = context.getProperty(PASSWORD).evaluateAttributeExpressions().getValue();
+
+        // Connection expects a non-null password, so set it to the empty string if it
+        // is not provided
+        if (password == null) {
+            password = "";
+        }
+
+        this.hasSnapshot = context.getProperty(SNAPSHOT).asBoolean();
+        this.includeBeginCommit = context.getProperty(INCLUDE_BEGIN_COMMIT).asBoolean();
+        this.initialLSN = context.getProperty(INIT_LSN).asLong() == null ? null : context.getProperty(INIT_LSN).asLong();
+
+        try {
+
+            // Ensure driverLocation and driverName are correct before establishing
+            // PostgreSQL connection to avoid failing after PostgreSQL messages are received.
+            // Actual JDBC connection is created after PostgreSQL client gets started,
+            // because we need the connect-able host same as the PostgreSQL client.
+            registerDriver(driverLocation, driverName);
+            this.connectionManager = new ConnectionManager();
+            this.connectionManager.setProperties(host, database, username, password, driverName);
+            this.connectionManager.createSQLConnection();
+            this.connectionManager.createReplicationConnection();
+
+            stateMap = stateManager.getState(Scope.CLUSTER);
+
+            this.pgEasyReplication = new PGEasyReplication(publicationName, replicationSlotName, this.connectionManager);
+            this.pgEasyReplication.initializeLogicalReplication(dropSlotIfExists);
+
+            if (stateMap.get(LAST_LSN_RECEIVED) != null) {
+                this.lastLSNReceived = Long.parseLong(stateMap.get(LAST_LSN_RECEIVED));
+                this.hasRun.set(true);
+            } else {
+                this.hasRun.set(false);
+            }
+
+        } catch (final Exception ioe) {
+            logger.error("Failed to retrieve values from the State Manager. Will not attempt " + "connection until this is accomplished.", ioe);
+            context.yield();
+            throw new ProcessException(ioe.getMessage(), ioe);
+        }
+    }
+
+    List<Event> getData() {
```

Comment:
Does it make sense here to get all available events, or up to some configurable limit for "Events per flow file"? The existing MySQL CDC processor was written before the record capability was added to NiFi, so it made more sense to split up events since most flows would have to split them up anyway for downstream processing. With the record-based components, would it be easier/better to have a larger number of events in each flow file? Or for downstream use cases is it still better to have 1-3 events per flow file?

Reply:
This method was included in onTrigger aiming for cleaner code. 

About events per flow file, I totally agree. Currently, each captured event becomes a flow file. So, with the record-based components we can improve this processor in several aspects. But, as you suggested, I believe it can be developed in a new JIRA.